nuxeo-runtime/nuxeo-runtime-server/src/main/java/org/nuxeo/runtime/server/ServerComponent.java

```diff
@@ -0,0 +1,126 @@
+import org.apache.commons.logging.LogFactory;
+import org.nuxeo.common.server.WebApplication;
+import org.nuxeo.runtime.RuntimeServiceException;
+import org.nuxeo.runtime.model.ComponentContext;
+import org.nuxeo.runtime.model.ComponentInstance;
+import org.nuxeo.runtime.model.DefaultComponent;
+
+/**
+ * This component registers and configures an embedded servlet container.
+ * <p>
+ * The actual implementation is delegated to a specific embedded server by implementing the interface
+ * {@link ServerConfigurator}.
+ *
+ * @since 10.2
+ */
+public class ServerComponent extends DefaultComponent {
+
+    private static final Log log = LogFactory.getLog(ServerComponent.class);
+
+    public static final String XP_WEB_APP = "webapp";
+
+    public static final String XP_SERVLET = "servlet";
+
+    public static final String XP_FILTER = "filter";
+
+    public static final String XP_LISTENER = "listener";
+
+    public static final String PORT_SYSTEM_PROP = "nuxeo.servlet-container.port";
+
+    protected static final String CONFIGURATOR_CLASS = "org.nuxeo.runtime.server.tomcat.TomcatServerConfigurator";
+
+    protected ServerConfigurator configurator;
+
+    protected int port;
+
+    @Override
+    public void activate(ComponentContext context) {
+        try {
+            configurator = (ServerConfigurator) Class.forName(CONFIGURATOR_CLASS).getConstructor().newInstance();
+        } catch (ReflectiveOperationException e) {
+            throw new RuntimeServiceException(e);
+        }
+        // config from test features
+        int p = -1;
+        String configPort = System.getProperty(PORT_SYSTEM_PROP);
+        if (StringUtils.isNotBlank(configPort)) {
+            try {
+                p = Integer.parseInt(configPort);
+            } catch (NumberFormatException e) {
+                log.error("Invalid port for embedded servlet container: " + configPort);
+            }
+        }
+        port = configurator.initialize(p);
+    }
+
+    @Override
+    public void deactivate(ComponentContext context) {
+        configurator.close();
+    }
+
+    @Override
+    public void registerContribution(Object contribution, String extensionPoint, ComponentInstance contributor) {
+        if (XP_WEB_APP.equals(extensionPoint)) {
+            configurator.addWepApp((WebApplication) contribution);
+        } else if (XP_FILTER.equals(extensionPoint)) {
+            configurator.addFilter((FilterDescriptor) contribution);
+        } else if (XP_SERVLET.equals(extensionPoint)) {
+            configurator.addServlet((ServletDescriptor) contribution);
+        } else if (XP_LISTENER.equals(extensionPoint)) {
+            configurator.addLifecycleListener((ServletContextListenerDescriptor) contribution);
+        }
+    }
+
+    @Override
+    public void unregisterContribution(Object contribution, String extensionPoint, ComponentInstance contributor) {
+        // we don't do anything special as this is a test component
```

Comment:
We won't have Deploy on method support. Is it wanted ?

Reply:
Let's say it's the same restriction that we had before with Jetty...